doc/book/box/indexes.rst

```diff
@@ -24,46 +20,454 @@ The first index defined on a space is called the **primary key index**,
+    :class: fact
+
+    The full information about creating an index is in section
+    :doc:`/reference/reference_lua/box_space/create_index`.
+
+--------------------------------------------------------------------------------
+Index types
+--------------------------------------------------------------------------------
+
+An index always has a **type**. Different types are intended for different
+usage scenarios.
+
+We give an overview of index features in the following table:
+
+..  container:: table
+
+    ..  list-table::
+        :header-rows: 1
+
+        *   - Feature
+            - TREE
+            - HASH
+            - RTREE
+            - BITSET
+
+        *   - unique
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - non-unique
+            - \+
+            - \-
+            - \+
+            - \+
+
+        *   - :ref:`is_nullable <box_space-is_nullable>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - multi-part
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - :ref:`multikey <box_space-path_multikey>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - :ref:`partial-key search <partial_key_search>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - can be primary key
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - exclude_null (version 2.8+)
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - :doc:`iterator types </reference/reference_lua/box_index/pairs>`
+            - ALL, EQ, REQ, GT, GE, LT, LE
+            - ALL, EQ, GT
+            - ALL, EQ, GT, GE, LT, LE, OVERLAPS, NEIGHBOR
+            - ALL, EQ, BITS_ALL_SET, BITS_ANY_SET, BITS_ALL_NOT_SET
+
+********************************************************************************
+TREE indexes
+********************************************************************************
+
+The default index type is 'TREE'.
+TREE indexes are provided by memtx and vinyl engines, can index unique and
+non-unique values, support partial key searches, comparisons and ordered results.
+
+This is a universal type of indexes, for most cases it will be the best choice.
+
+Additionally, memtx engine supports HASH, RTREE and BITSET indexes.
+
+********************************************************************************
+HASH indexes
+********************************************************************************
+
+HASH indexes require unique fields and loses to TREE in almost all respects.
+So we do not recommend to use it in the applications.
+HASH is now present in Tarantool mainly because of backward compatibility.
+
+Here are some tips. Do not use HASH index:
+
+* just if you want to
+* if you think that HASH is faster with no performance metering
+* if you want to iterate over the data
+* for primary key
```

Comment:
It would be great to describe why

Reply:
Doesn't the first paragraph describe it well?

Reply:
The first paragraph contradicts part below - `if you really need that 2-5% performance improvement` (that means that for some tasks it's more appropriate than TREE) and `if you save every byte on tuples` (to be honest, it looks like sneer) - yes, for some tasks it's important to save as much memory as possible.

Reply:
Ok, let's rewrite `if you really need that 2-5% performance improvement` as `if you measure, that it will give you accountable performance improvement`

> if you save every byte on tuples (to be honest, it looks like sneer) - yes, for some tasks it's important to save as much memory as possible.
See nothing wrong. In **that** case HASH is appropriate. "Normal" users should not use hash: it gives in future more problems, than gains.